src/scan/heap_reader_worker.cpp

```diff
@@ -0,0 +1,143 @@
+
+extern "C" {
+#include "postgres.h"
+#include "miscadmin.h"
+#include "access/heapam.h"
+#include "postmaster/bgworker.h"
+#include "storage/bufmgr.h"
+#include "storage/dsm.h"
+#include "tcop/tcopprot.h"
+}
+
+#include "pgduckdb/scan/heap_reader_worker.hpp"
+
+extern "C" {
+
+void
+PGDuckDBHeapReaderWorker(Datum arg) {
+	pqsignal(SIGTERM, die);
+	dsm_segment *thread_worker_segment = nullptr, *worker_shared_state_segment = nullptr;
+	BufferAccessStrategy strategy = GetAccessStrategy(BAS_BULKREAD);
+
+	BackgroundWorkerUnblockSignals();
+
+	pgduckdb::PostgresScanThreadWorker *thread_worker_shared_state = NULL;
+	pgduckdb::PostgresScanWorkerSharedState *worker_shared_state = NULL;
+
+	thread_worker_segment = dsm_attach(arg);
+	thread_worker_shared_state = (pgduckdb::PostgresScanThreadWorker *)dsm_segment_address(thread_worker_segment);
+
+	worker_shared_state_segment = dsm_attach(thread_worker_shared_state->workers_global_shared_state);
+	worker_shared_state = (pgduckdb::PostgresScanWorkerSharedState *)dsm_segment_address(worker_shared_state_segment);
+
+	BackgroundWorkerInitializeConnectionByOid(worker_shared_state->database, InvalidOid, 0);
+	StartTransactionCommand();
+
+	Relation relation = RelationIdGetRelation(worker_shared_state->relid);
+	bool thread_running = true;
+
+	while (thread_running) {
+
+		SpinLockAcquire(&worker_shared_state->worker_lock);
+
+		/* Nothing to scan next so finish this worker */
+		if (worker_shared_state->next_scan_start_block == worker_shared_state->n_blocks) {
+			SpinLockRelease(&worker_shared_state->worker_lock);
+			break;
+		}
+
+		uint32 worker_start_scan_block = worker_shared_state->next_scan_start_block;
+		uint32 worker_end_scan_block = worker_start_scan_block + pgduckdb::WORKER_SCAN_BLOCK;
+
+		if (worker_end_scan_block > worker_shared_state->n_blocks)
+			worker_end_scan_block = worker_shared_state->n_blocks;
+
+		worker_shared_state->next_scan_start_block = worker_end_scan_block;
+
+		SpinLockRelease(&worker_shared_state->worker_lock);
+
+		/* Let's fetch */
+		for (int i = worker_start_scan_block; i < worker_end_scan_block; i++) {
+
+			if (!thread_running) {
+				break;
+			}
+
+			Buffer buffer = ReadBufferExtended(relation, MAIN_FORKNUM, i, RBM_NORMAL, strategy);
+			LockBuffer(buffer, BUFFER_LOCK_SHARE);
+
+			/* is previous buffer done */
+			while (!pg_atomic_unlocked_test_flag(&thread_worker_shared_state->buffer_ready)) {
```

Comment:
Did you try adding a very small "sleep" in the loop rather than busy-spinning? I wonder if it would help performances.
Also could we use signals instead?
Should we have a timeout to handle situation where thread dies without ever resetting its `thread_running` or `buffer_ready` flags?